### PR TITLE
Add Nix flake

### PR DESCRIPTION
**Pull request by aceofscissors**
Dependencies can now be specified and automatically managed through Nix.

Since submodules are not used by default, you need to explicitly enable them in the flake URL or path. When building locally, use `nix build .?submodules=1`.

`nix develop` gives you a shell with all build dependencies and other tools. Building normally is still supported as well.

### CODE REVIEW

**Comment by xeniafox** [comment]
Can't look into it right now but will review soon

File: meson.options
Comment (xeniafox) [inline meson.options:1]: Why did you need to include GoogleTest?

Comment (aceofscissors) [reply meson.options:1]: ANTLR4 requires it as a dependency: https://github.com/neonredtech/astML/blob/d774f021ff2a3c3368060faf86442efa376095d7/subprojects/antlr4_CppRuntime/runtime/CMakeLists.txt#L71-L74

It is downloaded automatically during a normal build, however builds in Nix typically don't have a connection, and all required dependencies must be fetched before starting the build.

At the moment there is no generalized way to deal with this in Nix without manually passing [`FETCHCONTENT_SOURCE_DIR_*`](https://cmake.org/cmake/help/latest/module/FetchContent.html#variable:FETCHCONTENT_SOURCE_DIR_%3CuppercaseName%3E) to CMmake, although I have seen some projects with multiple CMake dependencies that use FetchContent scan the file looking for all occurrences of it.

Comment (xeniafox) [reply meson.options:1]: I see. Thanks for explaining it.

**Review by xeniafox — APPROVED** [review]
Looks good to me.
